File: web/lib/fleet/equipment-service.ts
import { getEquipmentExpirySummary } from "@/lib/fleet/expiry";
import type { EquipmentUpsertInput } from "@/lib/fleet/equipment-form-data";
import { requireOrgContext } from "@/lib/fleet/org-context";
import type {
  EquipmentAssignmentRecord,
  EquipmentDetail,
  EquipmentMaintenanceRecord,
  EquipmentRecord,
  EquipmentScheduleRecord,
  EquipmentType,
  EquipmentWithExpiry,
  VehicleInspectionRecord,
} from "@/lib/fleet/types";

export type EquipmentListFilters = {
  search?: string;
  type?: EquipmentType | "all";
  hasExpiryAlert?: boolean;
  inactiveOnly?: boolean;
  includeInactive?: boolean;
};

function normalizeNullableString(value: unknown): string | null {
  return typeof value === "string" ? value : null;
}

function normalizeNullableNumber(value: unknown): number | null {
  return typeof value === "number" ? value : null;
}

function crewNameFromJoin(value: unknown): string | null {
  if (Array.isArray(value)) {
    const first = value[0] as { name?: unknown } | undefined;
    return typeof first?.name === "string" ? first.name : null;
  }

  if (value && typeof value === "object" && "name" in value) {
    const name = (value as { name?: unknown }).name;
    return typeof name === "string" ? name : null;
  }

  return null;
}

function mapEquipmentRow(row: Record<string, unknown>): EquipmentWithExpiry {
  const equipment: EquipmentRecord = {
    id: String(row.id),
    org_id: String(row.org_id),
    name: String(row.name),
    type: row.type as EquipmentType,
    make: normalizeNullableString(row.make),
    model: normalizeNullableString(row.model),
    year: normalizeNullableNumber(row.year),
    vin_serial: normalizeNullableString(row.vin_serial),
    license_plate: normalizeNullableString(row.license_plate),
    dot_number: normalizeNullableString(row.dot_number),
    registration_expiry: normalizeNullableString(row.registration_expiry),
    insurance_expiry: normalizeNullableString(row.insurance_expiry),
    annual_inspection_due: normalizeNullableString(row.annual_inspection_due),
    is_shareable: Boolean(row.is_shareable),
    is_active: Boolean(row.is_active),
    notes: normalizeNullableString(row.notes),
    created_at: String(row.created_at),
    updated_at: String(row.updated_at),
  };

  return {
    ...equipment,
    expiry: getEquipmentExpirySummary(equipment),
  };
}

function mapAssignmentRow(row: Record<string, unknown>): EquipmentAssignmentRecord {
  return {
    id: String(row.id),
    org_id: String(row.org_id),
    equipment_id: String(row.equipment_id),
    crew_id: String(row.crew_id),
    assigned_date: String(row.assigned_date),
    notes: normalizeNullableString(row.notes),
    crew_name: crewNameFromJoin(row.crews),
    created_at: String(row.created_at),
    updated_at: String(row.updated_at),
  };
}

function mapScheduleRow(row: Record<string, unknown>): EquipmentScheduleRecord {
  return {
    id: String(row.id),
    org_id: String(row.org_id),
    equipment_id: String(row.equipment_id),
    crew_id: normalizeNullableString(row.crew_id),
    start_date: String(row.start_date),
    end_date: normalizeNullableString(row.end_date),
    notes: normalizeNullableString(row.notes),
    crew_name: crewNameFromJoin(row.crews),
    created_at: String(row.created_at),
    updated_at: String(row.updated_at),
  };
}

function mapMaintenanceRow(row: Record<string, unknown>): EquipmentMaintenanceRecord {
  return {
    id: String(row.id),
    org_id: String(row.org_id),
    equipment_id: String(row.equipment_id),
    type: String(row.type),
    performed_date: String(row.performed_date),
    mileage: normalizeNullableNumber(row.mileage),
    next_service_date: normalizeNullableString(row.next_service_date),
    next_service_mileage: normalizeNullableNumber(row.next_service_mileage),
    performed_by: normalizeNullableString(row.performed_by),
    cost: normalizeNullableNumber(row.cost),
    notes: normalizeNullableString(row.notes),
    receipt_url: normalizeNullableString(row.receipt_url),
    created_at: String(row.created_at),
    updated_at: String(row.updated_at),
  };
}

function driverNameFromJoin(value: unknown): string | null {
  if (!value || typeof value !== "object") {
    return null;
  }

  const employee = value as {
    display_name?: unknown;
    first_name?: unknown;
    last_name?: unknown;
  };

  if (typeof employee.display_name === "string" && employee.display_name.trim()) {
    return employee.display_name.trim();
  }

  const first = typeof employee.first_name === "string" ? employee.first_name.trim() : "";
  const last = typeof employee.last_name === "string" ? employee.last_name.trim() : "";
  const combined = `${first} ${last}`.trim();

  return combined || null;
}

function mapInspectionRow(row: Record<string, unknown>): VehicleInspectionRecord {
  const defects = Array.isArray(row.defects)
    ? row.defects.filter((item): item is string => typeof item === "string")
    : null;

  return {
    id: String(row.id),
    org_id: String(row.org_id),
    equipment_id: String(row.equipment_id),
    driver_id: String(row.driver_id),
    inspection_type: row.inspection_type as "pre_trip" | "post_trip",
    inspection_date: String(row.inspection_date),
    odometer: normalizeNullableNumber(row.odometer),
    passed: Boolean(row.passed),
    defects,
    driver_signature: normalizeNullableString(row.driver_signature),
    notes: normalizeNullableString(row.notes),
    driver_name: driverNameFromJoin(row.employees),
    created_at: String(row.created_at),
    updated_at: String(row.updated_at),
  };
}

function escapeForIlike(value: string): string {
  return value
    .replaceAll("%", "\\%")
    .replaceAll("_", "\\_")
    .replaceAll(",", " ")
    .trim();
}

function toNullable(value: string | null | undefined): string | null {
  if (!value) {
    return null;
  }

  const normalized = value.trim();
  return normalized.length > 0 ? normalized : null;
}

function toDateOnlyYmd(input: Date): string {
  return input.toISOString().split("T")[0];
}

export async function listEquipment(
  filters: EquipmentListFilters = {},
): Promise<EquipmentWithExpiry[]> {
  const { supabase, orgId } = await requireOrgContext();

  let query = supabase.from("equipment").select("*").eq("org_id", orgId);

  if (filters.type && filters.type !== "all") {
    query = query.eq("type", filters.type);
  }

  if (filters.inactiveOnly) {
    query = query.eq("is_active", false);
  } else if (!filters.includeInactive) {
    query = query.eq("is_active", true);
  }

  if (filters.search?.trim()) {
    const search = escapeForIlike(filters.search);
    query = query.or(
      `name.ilike.%${search}%,make.ilike.%${search}%,model.ilike.%${search}%,vin_serial.ilike.%${search}%,license_plate.ilike.%${search}%`,
    );
  }

  const { data, error } = await query.order("type", { ascending: true }).order("name", {
    ascending: true,
  });

  if (error) {
    throw new Error(error.message);
  }

  const rows = (data ?? []) as Record<string, unknown>[];
  const equipment = rows.map(mapEquipmentRow);

  if (!filters.hasExpiryAlert) {
    return equipment;
  }

  return equipment.filter((item) => item.expiry.hasAlert);
}

export async function getEquipmentById(
  equipmentId: string,
): Promise<EquipmentWithExpiry | null> {
  const { supabase, orgId } = await requireOrgContext();

  const { data, error } = await supabase
    .from("equipment")
    .select("*")
    .eq("org_id", orgId)
    .eq("id", equipmentId)
    .maybeSingle();

  if (error) {
    throw new Error(error.message);
  }

  if (!data) {
    return null;
  }

  return mapEquipmentRow(data as Record<string, unknown>);
}

export async function getEquipmentDetail(
  equipmentId: string,
): Promise<EquipmentDetail | null> {
  const { supabase, orgId } = await requireOrgContext();

  const equipment = await getEquipmentById(equipmentId);
  if (!equipment) {
    return null;
  }

  const { data: assignmentRows, error: assignmentError } = await supabase
    .from("equipment_assignments")
    .select("id,org_id,equipment_id,crew_id,assigned_date,notes,created_at,updated_at,crews(name)")
    .eq("org_id", orgId)
    .eq("equipment_id", equipmentId)
    .order("assigned_date", { ascending: false });

  if (assignmentError) {
    throw new Error(assignmentError.message);
  }

  const { data: scheduleRows, error: scheduleError } = await supabase
    .from("equipment_schedule")
    .select("id,org_id,equipment_id,crew_id,start_date,end_date,notes,created_at,updated_at,crews(name)")
    .eq("org_id", orgId)
    .eq("equipment_id", equipmentId)
    .order("start_date", { ascending: true });

  if (scheduleError) {
    throw new Error(scheduleError.message);
  }

  const { data: maintenanceRows, error: maintenanceError } = await supabase
    .from("equipment_maintenance")
    .select("*")
    .eq("org_id", orgId)
    .eq("equipment_id", equipmentId)
    .order("performed_date", { ascending: false });

  if (maintenanceError) {
    throw new Error(maintenanceError.message);
  }

  const { data: inspectionRows, error: inspectionError } = await supabase
    .from("vehicle_inspections")
    .select("*,employees(first_name,last_name,display_name)")
    .eq("org_id", orgId)
    .eq("equipment_id", equipmentId)
    .order("inspection_date", { ascending: false });

  if (inspectionError) {
    throw new Error(inspectionError.message);
  }

  const assignments = ((assignmentRows ?? []) as Record<string, unknown>[]).map(mapAssignmentRow);
  const schedule = ((scheduleRows ?? []) as Record<string, unknown>[]).map(mapScheduleRow);
  const maintenance = ((maintenanceRows ?? []) as Record<string, unknown>[]).map(mapMaintenanceRow);
  const inspections = ((inspectionRows ?? []) as Record<string, unknown>[]).map(mapInspectionRow);

  return {
    equipment,
    assignments,
    currentAssignment: assignments[0] ?? null,
    schedule,
    maintenance,
    inspections,
  };
}

export async function createEquipment(
  input: EquipmentUpsertInput,
): Promise<EquipmentWithExpiry> {
  const { supabase, orgId } = await requireOrgContext();

  const { data, error } = await supabase
    .from("equipment")
    .insert({
      org_id: orgId,
      name: input.name,
      type: input.type,
      make: input.make,
      model: input.model,
      year: input.year,
      vin_serial: input.vin_serial,
      license_plate: input.license_plate,
      dot_number: input.dot_number,
      registration_expiry: input.registration_expiry,
      insurance_expiry: input.insurance_expiry,
      annual_inspection_due: input.annual_inspection_due,
      is_shareable: input.is_shareable,
      is_active: input.is_active,
      notes: input.notes,
    })
    .select("*")
    .single();

  if (error) {
    throw new Error(error.message);
  }

  return mapEquipmentRow(data as Record<string, unknown>);
}

export async function updateEquipment(
  equipmentId: string,
  input: EquipmentUpsertInput,
): Promise<EquipmentWithExpiry> {
  const { supabase, orgId } = await requireOrgContext();

  const { data, error } = await supabase
    .from("equipment")
    .update({
      name: input.name,
      type: input.type,
      make: input.make,
      model: input.model,
      year: input.year,
      vin_serial: input.vin_serial,
      license_plate: input.license_plate,
      dot_number: input.dot_number,
      registration_expiry: input.registration_expiry,
      insurance_expiry: input.insurance_expiry,
      annual_inspection_due: input.annual_inspection_due,
      is_shareable: input.is_shareable,
      is_active: input.is_active,
      notes: input.notes,
    })
    .eq("org_id", orgId)
    .eq("id", equipmentId)
    .select("*")
    .single();

  if (error) {
    throw new Error(error.message);
  }

  return mapEquipmentRow(data as Record<string, unknown>);
}

export async function checkEquipmentAvailability(input: {
  equipmentId: string;
  startDate: string;
  endDate?: string | null;
  excludeScheduleId?: string | null;
}): Promise<boolean> {
  const { supabase, orgId } = await requireOrgContext();

  const { data: equipmentRow, error: equipmentError } = await supabase
    .from("equipment")
    .select("is_active,is_shareable")
    .eq("org_id", orgId)
    .eq("id", input.equipmentId)
    .maybeSingle();

  if (equipmentError) {
    throw new Error(equipmentError.message);
  }

  if (!equipmentRow || equipmentRow.is_active === false) {
    return false;
  }

  if (equipmentRow.is_shareable === true) {
    return true;
  }

  const normalizedEndDate = toNullable(input.endDate) ?? input.startDate;

  const { data: rpcConflict, error: rpcError } = await supabase.rpc(
    "has_equipment_schedule_conflict",
    {
      check_org_id: orgId,
      check_equipment_id: input.equipmentId,
      check_start_date: input.startDate,
      check_end_date: normalizedEndDate,
      exclude_schedule_id: toNullable(input.excludeScheduleId),
    },
  );

  if (!rpcError && typeof rpcConflict === "boolean") {
    return !rpcConflict;
  }

  let overlapQuery = supabase
    .from("equipment_schedule")
    .select("id")
    .eq("org_id", orgId)
    .eq("equipment_id", input.equipmentId)
    .lte("start_date", normalizedEndDate)
    .or(`end_date.is.null,end_date.gte.${input.startDate}`);

  const excludedId = toNullable(input.excludeScheduleId);
  if (excludedId) {
    overlapQuery = overlapQuery.neq("id", excludedId);
  }

  const { data: overlaps, error: overlapError } = await overlapQuery;

  if (overlapError) {
    throw new Error(overlapError.message);
  }

  return (overlaps ?? []).length === 0;
}

export function formatDateRange(startDate: string, endDate: string | null): string {
  if (!endDate || endDate === startDate) {
    return startDate;
  }

  return `${startDate} to ${endDate}`;
}

export function todayYmd(): string {
  return toDateOnlyYmd(new Date());
}

File: lib/features/fleet/equipment_service.dart
import 'dart:typed_data';

import 'package:riverpod_annotation/riverpod_annotation.dart';
import 'package:supabase_flutter/supabase_flutter.dart';

import '../../core/network/supabase_client_provider.dart';
import 'models/fleet_models.dart';

part 'equipment_service.g.dart';

class EquipmentUpsertRequest {
  const EquipmentUpsertRequest({
    required this.name,
    required this.type,
    this.make,
    this.model,
    this.year,
    this.vinSerial,
    this.licensePlate,
    this.dotNumber,
    this.registrationExpiry,
    this.insuranceExpiry,
    this.annualInspectionDue,
    this.isShareable = false,
    this.isActive = true,
    this.notes,
  });

  final String name;
  final String type;
  final String? make;
  final String? model;
  final int? year;
  final String? vinSerial;
  final String? licensePlate;
  final String? dotNumber;
  final DateTime? registrationExpiry;
  final DateTime? insuranceExpiry;
  final DateTime? annualInspectionDue;
  final bool isShareable;
  final bool isActive;
  final String? notes;
}

class EquipmentRequestCreate {
  const EquipmentRequestCreate({
    required this.equipmentId,
    required this.crewId,
    required this.requestedBy,
    required this.startDate,
    this.endDate,
    this.jobId,
    this.notes,
  });

  final String equipmentId;
  final String crewId;
  final String? requestedBy;
  final DateTime startDate;
  final DateTime? endDate;
  final String? jobId;
  final String? notes;
}

class EquipmentService {
  EquipmentService(this._client);

  final SupabaseClient _client;

  Future<List<Equipment>> listEquipment({
    required String orgId,
    String search = '',
    String type = 'all',
    bool includeInactive = true,
    bool hasExpiryAlert = false,
    bool inactiveOnly = false,
  }) async {
    dynamic query = _client.from('equipment').select('*').eq('org_id', orgId);

    if (type != 'all') {
      query = query.eq('type', type);
    }

    if (inactiveOnly) {
      query = query.eq('is_active', false);
    } else if (!includeInactive) {
      query = query.eq('is_active', true);
    }

    final normalizedSearch = search.trim();
    if (normalizedSearch.isNotEmpty) {
      final escaped = _escapeForLike(normalizedSearch);
      query = query.or(
        'name.ilike.%$escaped%,make.ilike.%$escaped%,model.ilike.%$escaped%,vin_serial.ilike.%$escaped%,license_plate.ilike.%$escaped%',
      );
    }

    final rows = await query
        .order('type', ascending: true)
        .order('name', ascending: true);

    final equipment = (rows as List)
        .cast<Map<String, dynamic>>()
        .map(Equipment.fromJson)
        .toList();

    if (!hasExpiryAlert) {
      return equipment;
    }

    return equipment.where(_hasExpiryAlert).toList();
  }

  Future<EquipmentDetail?> getEquipmentDetail({
    required String orgId,
    required String equipmentId,
  }) async {
    final equipmentRow = await _client
        .from('equipment')
        .select('*')
        .eq('org_id', orgId)
        .eq('id', equipmentId)
        .maybeSingle();

    if (equipmentRow == null) {
      return null;
    }

    final assignmentRows = await _client
        .from('equipment_assignments')
        .select('*, crews(name)')
        .eq('org_id', orgId)
        .eq('equipment_id', equipmentId)
        .order('assigned_date', ascending: false);

    final assignments = (assignmentRows as List)
        .cast<Map<String, dynamic>>()
        .map(
          (row) => EquipmentAssignment.fromJson({
            ...row,
            'crew_name': (row['crews'] as Map<String, dynamic>?)?['name'],
          }),
        )
        .toList();

    final today = _dateOnly(DateTime.now());
    final scheduleRows = await _client
        .from('equipment_schedule')
        .select('*, crews(name)')
        .eq('org_id', orgId)
        .eq('equipment_id', equipmentId)
        .gte('start_date', _asYmd(today))
        .order('start_date', ascending: true);

    final schedule = (scheduleRows as List)
        .cast<Map<String, dynamic>>()
        .map(
          (row) => EquipmentSchedule.fromJson({
            ...row,
            'crew_name': (row['crews'] as Map<String, dynamic>?)?['name'],
          }),
        )
        .toList();

    final maintenanceRows = await _client
        .from('equipment_maintenance')
        .select('*')
        .eq('org_id', orgId)
        .eq('equipment_id', equipmentId)
        .order('performed_date', ascending: false);

    final maintenance = (maintenanceRows as List)
        .cast<Map<String, dynamic>>()
        .map(EquipmentMaintenance.fromJson)
        .toList();

    final inspectionRows = await _client
        .from('vehicle_inspections')
        .select('*, employees(first_name,last_name,display_name)')
        .eq('org_id', orgId)
        .eq('equipment_id', equipmentId)
        .order('inspection_date', ascending: false);

    final inspections = (inspectionRows as List)
        .cast<Map<String, dynamic>>()
        .map(
          (row) => VehicleInspection.fromJson({
            ...row,
            'driver_name': _resolvedEmployeeName(
              row['employees'] as Map<String, dynamic>?,
            ),
          }),
        )
        .toList();

    return EquipmentDetail(
      equipment: Equipment.fromJson(equipmentRow),
      assignments: assignments,
      currentAssignment: assignments.isEmpty ? null : assignments.first,
      schedule: schedule,
      maintenance: maintenance,
      inspections: inspections,
    );
  }

  Future<Equipment> createEquipment({
    required String orgId,
    required EquipmentUpsertRequest request,
  }) async {
    final row = await _client
        .from('equipment')
        .insert(_equipmentWriteMap(orgId: orgId, request: request))
        .select('*')
        .single();

    return Equipment.fromJson(row);
  }

  Future<Equipment> updateEquipment({
    required String orgId,
    required String equipmentId,
    required EquipmentUpsertRequest request,
  }) async {
    final row = await _client
        .from('equipment')
        .update(_equipmentWriteMap(orgId: orgId, request: request))
        .eq('org_id', orgId)
        .eq('id', equipmentId)
        .select('*')
        .single();

    return Equipment.fromJson(row);
  }

  Future<bool> checkAvailability({
    required String orgId,
    required String equipmentId,
    required DateTime startDate,
    DateTime? endDate,
    String? excludeScheduleId,
  }) async {
    final row = await _client
        .from('equipment')
        .select('is_active,is_shareable')
        .eq('org_id', orgId)
        .eq('id', equipmentId)
        .maybeSingle();

    if (row == null) {
      return false;
    }

    final isActive = row['is_active'] as bool? ?? false;
    if (!isActive) {
      return false;
    }

    final isShareable = row['is_shareable'] as bool? ?? false;
    if (isShareable) {
      return true;
    }

    final start = _dateOnly(startDate);
    final end = _dateOnly(endDate ?? startDate);

    try {
      final conflict = await _client.rpc(
        'has_equipment_schedule_conflict',
        params: {
          'check_org_id': orgId,
          'check_equipment_id': equipmentId,
          'check_start_date': _asYmd(start),
          'check_end_date': _asYmd(end),
          'exclude_schedule_id': excludeScheduleId,
        },
      );

      if (conflict is bool) {
        return !conflict;
      }
    } catch (_) {
      // Fall back to direct overlap query if RPC is unavailable.
    }

    dynamic query = _client
        .from('equipment_schedule')
        .select('id')
        .eq('org_id', orgId)
        .eq('equipment_id', equipmentId)
        .lte('start_date', _asYmd(end))
        .or('end_date.is.null,end_date.gte.${_asYmd(start)}');

    if (excludeScheduleId != null && excludeScheduleId.isNotEmpty) {
      query = query.neq('id', excludeScheduleId);
    }

    final overlaps = await query;
    return (overlaps as List).isEmpty;
  }

  String getExpiryStatus(DateTime? date, {DateTime? today}) {
    if (date == null) {
      return 'ok';
    }

    final now = _dateOnly(today ?? DateTime.now());
    final expiryDate = _dateOnly(date);
    final daysUntilExpiry = expiryDate.difference(now).inDays;

    if (daysUntilExpiry < 0) {
      return 'expired';
    }

    if (daysUntilExpiry <= 30) {
      return 'critical';
    }

    if (daysUntilExpiry <= 60) {
      return 'warning';
    }

    return 'ok';
  }

  Future<EquipmentRequest> submitEquipmentRequest({
    required String orgId,
    required EquipmentRequestCreate request,
  }) async {
    final inserted = await _client
        .from('equipment_requests')
        .insert({
          'org_id': orgId,
          'equipment_id': request.equipmentId,
          'crew_id': request.crewId,
          'requested_by': request.requestedBy,
          'start_date': _asYmd(request.startDate),
          'end_date': request.endDate == null ? null : _asYmd(request.endDate!),
          'job_id': _nullIfBlank(request.jobId),
          'notes': _nullIfBlank(request.notes),
          'status': 'pending',
        })
        .select('*')
        .single();

    final requestModel = EquipmentRequest.fromJson(inserted);

    final settingsRow = await _client
        .from('org_settings')
        .select('equipment_approval_required')
        .eq('org_id', orgId)
        .maybeSingle();

    final approvalRequired =
        settingsRow?['equipment_approval_required'] as bool? ?? false;

    if (!approvalRequired) {
      return approveEquipmentRequest(
        orgId: orgId,
        requestId: requestModel.id,
        reviewedBy: request.requestedBy,
      );
    }

    return requestModel;
  }

  Future<EquipmentRequest> approveEquipmentRequest({
    required String orgId,
    required String requestId,
    String? reviewedBy,
  }) async {
    final updated = await _client
        .from('equipment_requests')
        .update({
          'status': 'approved',
          'reviewed_by': reviewedBy,
          'reviewed_at': DateTime.now().toUtc().toIso8601String(),
        })
        .eq('org_id', orgId)
        .eq('id', requestId)
        .select('*')
        .single();

    final request = EquipmentRequest.fromJson(updated);

    await _client.from('equipment_schedule').insert({
      'org_id': orgId,
      'equipment_id': request.equipmentId,
      'crew_id': request.crewId,
      'job_id': request.jobId,
      'start_date': _asYmd(request.startDate),
      'end_date': request.endDate == null ? null : _asYmd(request.endDate!),
      'notes': request.notes,
    });

    return request;
  }

  Future<EquipmentRequest> denyEquipmentRequest({
    required String orgId,
    required String requestId,
    required String reason,
    String? reviewedBy,
  }) async {
    final updated = await _client
        .from('equipment_requests')
        .update({
          'status': 'denied',
          'reviewed_by': reviewedBy,
          'reviewed_at': DateTime.now().toUtc().toIso8601String(),
          'notes': reason.trim(),
        })
        .eq('org_id', orgId)
        .eq('id', requestId)
        .select('*')
        .single();

    return EquipmentRequest.fromJson(updated);
  }

  bool _hasExpiryAlert(Equipment equipment) {
    final registration = getExpiryStatus(equipment.registrationExpiry);
    final insurance = getExpiryStatus(equipment.insuranceExpiry);
    final inspection = getExpiryStatus(equipment.annualInspectionDue);

    return registration != 'ok' || insurance != 'ok' || inspection != 'ok';
  }

  Map<String, dynamic> _equipmentWriteMap({
    required String orgId,
    required EquipmentUpsertRequest request,
  }) {
    return {
      'org_id': orgId,
      'name': request.name.trim(),
      'type': request.type,
      'make': _nullIfBlank(request.make),
      'model': _nullIfBlank(request.model),
      'year': request.year,
      'vin_serial': _nullIfBlank(request.vinSerial),
      'license_plate': _nullIfBlank(request.licensePlate),
      'dot_number': _nullIfBlank(request.dotNumber),
      'registration_expiry':
          request.registrationExpiry == null ? null : _asYmd(request.registrationExpiry!),
      'insurance_expiry':
          request.insuranceExpiry == null ? null : _asYmd(request.insuranceExpiry!),
      'annual_inspection_due': request.annualInspectionDue == null
          ? null
          : _asYmd(request.annualInspectionDue!),
      'is_shareable': request.isShareable,
      'is_active': request.isActive,
      'notes': _nullIfBlank(request.notes),
    };
  }

  String _resolvedEmployeeName(Map<String, dynamic>? employeeRow) {
    if (employeeRow == null) {
      return 'Unknown Driver';
    }

    final displayName = (employeeRow['display_name'] as String?)?.trim();
    if (displayName != null && displayName.isNotEmpty) {
      return displayName;
    }

    final firstName = (employeeRow['first_name'] as String?)?.trim() ?? '';
    final lastName = (employeeRow['last_name'] as String?)?.trim() ?? '';
    final fullName = '$firstName $lastName'.trim();
    if (fullName.isNotEmpty) {
      return fullName;
    }

    return 'Unknown Driver';
  }

  String? _nullIfBlank(String? value) {
    final trimmed = value?.trim();
    if (trimmed == null || trimmed.isEmpty) {
      return null;
    }
    return trimmed;
  }

  DateTime _dateOnly(DateTime value) {
    return DateTime(value.year, value.month, value.day);
  }

  String _asYmd(DateTime value) {
    return value.toIso8601String().split('T').first;
  }

  String _escapeForLike(String value) {
    return value.replaceAll('%', r'\%').replaceAll(',', ' ');
  }
}

@riverpod
EquipmentService equipmentService(Ref ref) {
  return EquipmentService(ref.read(supabaseClientProvider));
}

class MaintenanceService {
  MaintenanceService(this._client);

  final SupabaseClient _client;

  Future<EquipmentMaintenance> logMaintenance({
    required String orgId,
    required String equipmentId,
    required String type,
    required DateTime performedDate,
    int? mileage,
    DateTime? nextServiceDate,
    int? nextServiceMileage,
    String? performedBy,
    double? cost,
    String? notes,
    Uint8List? receiptBytes,
    String? receiptFileName,
    String? existingReceiptUrl,
  }) async {
    String? receiptUrl = existingReceiptUrl;

    if (receiptBytes != null && receiptBytes.isNotEmpty) {
      final fileName = receiptFileName ??
          'receipt-${DateTime.now().millisecondsSinceEpoch}.bin';
      final path = '$orgId/equipment/$equipmentId/maintenance/$fileName';

      await _client.storage.from('fleet-docs').uploadBinary(
            path,
            receiptBytes,
            fileOptions: const FileOptions(upsert: true),
          );
      receiptUrl = _client.storage.from('fleet-docs').getPublicUrl(path);
    }

    final row = await _client
        .from('equipment_maintenance')
        .insert({
          'org_id': orgId,
          'equipment_id': equipmentId,
          'type': type.trim(),
          'performed_date': _asYmd(performedDate),
          'mileage': mileage,
          'next_service_date':
              nextServiceDate == null ? null : _asYmd(nextServiceDate),
          'next_service_mileage': nextServiceMileage,
          'performed_by': _nullIfBlank(performedBy),
          'cost': cost,
          'notes': _nullIfBlank(notes),
          'receipt_url': _nullIfBlank(receiptUrl),
        })
        .select('*')
        .single();

    return EquipmentMaintenance.fromJson(row);
  }

  String _asYmd(DateTime value) {
    return value.toIso8601String().split('T').first;
  }

  String? _nullIfBlank(String? value) {
    final trimmed = value?.trim();
    if (trimmed == null || trimmed.isEmpty) {
      return null;
    }
    return trimmed;
  }
}

@riverpod
MaintenanceService maintenanceService(Ref ref) {
  return MaintenanceService(ref.read(supabaseClientProvider));
}

class DvirService {
  DvirService(this._client);

  final SupabaseClient _client;

  Future<List<Map<String, String>>> listDriverOptions({
    required String orgId,
  }) async {
    final rows = await _client
        .from('employees')
        .select('id,first_name,last_name,display_name,employee_status')
        .eq('org_id', orgId)
        .eq('employee_status', 'active')
        .order('last_name', ascending: true)
        .order('first_name', ascending: true);

    return (rows as List)
        .cast<Map<String, dynamic>>()
        .map(
          (row) => {
            'id': row['id'] as String,
            'name': _resolvedEmployeeName(row),
          },
        )
        .toList();
  }

  Future<VehicleInspection> logInspection({
    required String orgId,
    required String equipmentId,
    required String driverId,
    required String inspectionType,
    required DateTime inspectionDate,
    required bool passed,
    int? odometer,
    List<String>? defects,
    String? driverSignature,
    Uint8List? signatureBytes,
    String? signatureFileName,
    String? notes,
  }) async {
    String? signatureUrl = _nullIfBlank(driverSignature);

    if (signatureBytes != null && signatureBytes.isNotEmpty) {
      final fileName = signatureFileName ??
          'signature-${DateTime.now().millisecondsSinceEpoch}.bin';
      final path = '$orgId/equipment/$equipmentId/dvir/$fileName';

      await _client.storage.from('fleet-docs').uploadBinary(
            path,
            signatureBytes,
            fileOptions: const FileOptions(upsert: true),
          );
      signatureUrl = _client.storage.from('fleet-docs').getPublicUrl(path);
    }

    final row = await _client
        .from('vehicle_inspections')
        .insert({
          'org_id': orgId,
          'equipment_id': equipmentId,
          'driver_id': driverId,
          'inspection_type': inspectionType,
          'inspection_date': _asYmd(inspectionDate),
          'odometer': odometer,
          'passed': passed,
          'defects': passed ? null : defects,
          'driver_signature': signatureUrl,
          'notes': _nullIfBlank(notes),
        })
        .select('*')
        .single();

    return VehicleInspection.fromJson(row);
  }

  String _asYmd(DateTime value) {
    return value.toIso8601String().split('T').first;
  }

  String _resolvedEmployeeName(Map<String, dynamic> employeeRow) {
    final displayName = (employeeRow['display_name'] as String?)?.trim();
    if (displayName != null && displayName.isNotEmpty) {
      return displayName;
    }

    final firstName = (employeeRow['first_name'] as String?)?.trim() ?? '';
    final lastName = (employeeRow['last_name'] as String?)?.trim() ?? '';
    final fullName = '$firstName $lastName'.trim();
    if (fullName.isNotEmpty) {
      return fullName;
    }

    return 'Unknown Driver';
  }

  String? _nullIfBlank(String? value) {
    final trimmed = value?.trim();
    if (trimmed == null || trimmed.isEmpty) {
      return null;
    }
    return trimmed;
  }
}

@riverpod
DvirService dvirService(Ref ref) {
  return DvirService(ref.read(supabaseClientProvider));
}
